src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    /// <summary>
+    /// Pipe implementation of a reader for HTTP Multipart Content Type
+    /// </summary>
+    public class MultipartPipeReader
+    {
+        private const int StackAllocThreshold = 128;
+
+        private readonly PipeReader _pipeReader;
+        private readonly MultipartBoundary _boundary;
+        private MultipartSectionPipeReader? _currentSectionReader;
+
+        private static ReadOnlySpan<byte> ColonDelimiter => new byte[] { (byte)':' };
+        private static ReadOnlySpan<byte> CrlfDelimiter => new byte[] { (byte)'\r', (byte)'\n' };
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="boundary">the Multipart boundary</param>
+        /// <param name="pipeReader">a Pipe Reader</param>
+        public MultipartPipeReader(string boundary, PipeReader pipeReader)
+        {
+            _pipeReader = pipeReader ?? throw new ArgumentNullException(nameof(pipeReader));
+            _boundary = new MultipartBoundary(boundary ?? throw new ArgumentNullException(nameof(boundary)), false);
+        }
+
+        /// <summary>
+        /// The limit for the number of headers to read.
+        /// </summary>
+        public int HeadersCountLimit { get; set; } = MultipartReader.DefaultHeadersCountLimit;
+
+        /// <summary>
+        /// The combined size limit for headers per multipart section.
+        /// </summary>
+        public int HeadersLengthLimit { get; set; } = MultipartReader.DefaultHeadersLengthLimit;
+
+        /// <summary>
+        /// The optional size limit for each section body.
+        /// </summary>
+        public long? BodyLengthLimit { get; set; }
+
+        /// <summary>
+        /// Read the next Multipart Section
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        public async Task<MultipartPipeSection?> ReadNextSectionAsync(CancellationToken cancellationToken = default)
+        {
+            if (_currentSectionReader == null) // first call.
+            {
+                // This section will drain any preamble data and remove the first boundary marker.
+                _currentSectionReader = new MultipartSectionPipeReader(_pipeReader, _boundary) { LengthLimit = HeadersLengthLimit };
+            }
+
+            await _currentSectionReader.DrainAsync(cancellationToken);
+
+            if (_currentSectionReader.FinalBoundaryFound)
+            {
+                return null;
```

Comment:
In the pipeReader implementation we are reading to the end of the pipeReader once we find the finalBoundary, so in practice the behavior is the same, and there is no need to drain the trailing data - basically the difference is that the pipe is just drained to end as soon as the final boundary is found , instead of waiting for a subsequent DrainAsync.

https://github.com/dotnet/aspnetcore/blob/b784d8ea2d34513dd12a1f533a17431679c22c63/src/Http/WebUtilities/src/MultipartSectionPipeReader.cs#L187-L195